### PR TITLE
Compilation & EVM integration

### PR DESCRIPTION
**Pull request by tomusdrw**

### CODE REVIEW

File: solaris/src/evm.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        }.fake_sign(sender);
+
+        self.evm.transact(&env_info, transaction, ethcore::trace::NoopTracer, ethcore::trace::NoopVMTracer);
+
+        self
+    }
+
+    pub fn logs(&self, _filter: ::ethabi::TopicFilter) -> Vec<()> {
+        // TODO [ToDr] Add filter querying
+        self.logs.iter().map(|_| ()).collect()
```
Comment (lght) [inline solaris/src/evm.rs]: Is this just acting as a stub for now?

File: solc/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "solc"
```
Comment (lght) [inline solc/Cargo.toml]: Is this going to be a wrapper around solc, or are we planning to port solc to Rust?

**Comment by lght** [comment]
Just ran into a build error:

```
   Compiling bitflags v1.0.1
error: expected ident, found #
   --> /.cargo/registry/src/github.com-1ecc6299db9ec823/bitflags-1.0.1/src/lib.rs:423:29
    |
423 |                               #[allow(deprecated)]
    |                               ^
    |
   ::: /.cargo/registry/src/github.com-1ecc6299db9ec823/bitflags-1.0.1/src/example_generated.rs
    |
4   | / bitflags! {
5   | |     /// This is the same `Flags` struct defined in the [crate level example](../index.html#example).
6   | |     /// Note that this struct is just for documentation purposes only, it must not be used outside
7   | |     /// this crate.
...   |
13  | |     }
14  | | }
    | |_- in this macro invocation
```

Was able to fix the error by removing the example referenced in the error. Also noticed two versions of `bitflags` being used, is this intentional?

File: example/src/main.rs [hunk 3]
```diff
@@ -3,30 +3,82 @@ extern crate ethabi;
...
+
+    (evm, contract)
+}
+
+#[test]
+fn badge_reg_test_fee() {
+    let (mut evm, contract) = setup();
+
+    // Initial fee is 1 ETH
+    assert_eq!(unit::convert(contract.functions().fee().call(&mut evm).unwrap()), unit::ether(1));
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply example/src/main.rs]: Yes, that's a good idea. Just got the idea when I started writing the second test and didn't refactor this one.

File: solaris/src/evm.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            gas: convert_u256(self.gas),
+            action: transaction::Action::Create,
+            value: convert_u256(self.value),
+            data: code.to_vec(),
+        }.fake_sign((&*self.sender).into());
+
+        let tracers = self.tracers();
+        match self.evm.transact(&env_info, transaction, tracers.0, tracers.1) {
+            TransactResult::Ok { contract_address, .. } => {
+                self.contract_address = contract_address.map(|x| (&*x).into());
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply solaris/src/evm.rs]: The problem is that `ethcore` crate and `solaris` use different types for primitives like U256 and H160.

We use `ethereum-types` (not released on crates.io yet, but will be soon), which is a new take on this. `ethcore` still relies on old `ethcore-bigint` stuff. The code is identical, just the way the types are created was reworked by @debris.

I did this in the anticipation of parity migration to `ethereum-types` as well. And it won't panic, it just converts from `H160 -> [u8;20] -> H160`

File: solaris/src/evm.rs (hunk 1 above)
Comment (tomusdrw) [reply solaris/src/evm.rs]: Yes, should do actual filtering by topics from `ethabi::TopicFilter`.

File: solaris/src/evm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lght) [reply solaris/src/evm.rs]: > I did this in the anticipation of parity migration to ethereum-types as well. And it won't panic, it just converts from H160 -> [u8;20] -> H160

Got it, thought you might be doing something like this. Need to remember this is Rust, and stop getting freaked out by pointers. This part looks good to me.

File: example/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lght) [reply example/src/main.rs]: Very nice. Once this is updated, everything is golden.

**Comment by axelchalon** [comment]
really good stuff!

just something I thought (might be nitpicky): I find the syntax a bit counter-intuitive (also due to the current ethabi syntax). For example in:
```
reg.register().transact(sol::address(10), sol::bytes32("test"),
         evm
         .with_value(unit::ether(2))
         .with_sender(5.into())
         .ensure_funds()
         )?;
```
we're calling functions on `reg`, the contract encoder/decoder helper, but it's the CALLER (evm) passed as parameter that is actually doing the main job.

Don't you think that making the evm function call go first would be a good idea? Not yet sure how to implement this; it would be \*roughly\* something in this direction: (maybe using macros somehow)
```
evm
    .with_value(unit::ether(2))
    .with_sender(5.into())
    .ensure_funds()
    .transact2(reg.register(), sol::address(10), sol::bytes32("test"))?;
    
fn transact2<...>(caller: CALLER, cfn: ContractFunction, inputs...) {
/* ... */
}

// On the other hand this order of argument doesn't make sense
transact2(|input_bytes| { /* interact with blockchain */ }, reg.register(), input1, input2)
// while this one does (input gets encoded, then callback gets called)
transact2(reg.register(), input1, input2, |input_bytes| { /* interact with blockchain */ })
```

**Comment by tomusdrw** [comment]
@axelchalon totally agree wrt to API, have a look at the proposal of changes to ethabi:
https://gist.github.com/tomusdrw/e8b6b3fa448a9935b35858e1aa6ceaa9

This would allow use to do:
```rust
evm.transact(|contract| contract.register(sol::address(10), sol::bytes32("test"))?;
```

**Pull request merged**